### PR TITLE
allow `scope` structs referring to themselves

### PR DESCRIPTION
**Pull request by aG0aep6G**
This is a follow-up to #10773. Taking the address of the value should not prevent `scope` inference.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
Do you want to add an issue or changelog to this?

**Comment by aG0aep6G** [comment]
> Do you want to add an issue or changelog to this?

I consider this a fix-up of the last one. I don't think it needs another entry.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -11926,16 +11926,20 @@ private bool checkAddressVar(Scope* sc, UnaExp exp, VarDeclaration v)
...
             const(char)* p = v.isParameter() ? "parameter" : "local";
+            const bool isRef = (v.storage_class & (STC.ref_ | STC.out_)) != 0;
             if (global.params.vsafe)
             {
-                // Taking the address of v means it cannot be set to 'scope' later
-                v.storage_class &= ~STC.maybescope;
-                v.doNotInferScope = true;
-                if (exp.e1.type.hasPointers() && v.storage_class & STC.scope_ &&
-                    !(sc.flags & SCOPE.debug_) && sc.func.setUnsafe())
+                if (exp.e1.type.hasPointers() || exp.isDelegateExp() || isRef)
```
Comment (aG0aep6G) [inline src/dmd/expressionsem.d]: I'm not at all sure that this condition is sufficient. I've just reacted to failing tests.

Comment (thewilsonator) [reply src/dmd/expressionsem.d]: Hmm, did you test with delegates, AAs and classes in your last PR?

Comment (aG0aep6G) [reply src/dmd/expressionsem.d]: > Hmm, did you test with delegates, AAs and classes in your last PR?

I don't know if I tested with delegates and AAs before, but I did just now. As types, they're caught by `hasPointers` (as expected).

I also just re-ran the test suite without the `exp.isDelegateExp()` check, and it passes. I guess the corresponding test (IIRC: <https://github.com/dlang/dmd/blob/25e9ef8968c5da854b4ff1bd75e2c63893ac446e/test/fail_compilation/fix18575.d#L23>) also features a `ref` parameter, and the `DelegateExp` was a red herring to me. I can't come up with a new test that would need `exp.isDelegateExp()`, either.

So maybe `exp.e1.type.hasPointers() || isRef` is sufficient. But to make things worse, if the test line from above is changed from a direct `return` to this:

```d
auto ghi(ref S s) { auto dg = &s.bar; return dg; }
```

DMD doesn't catch it anymore. So there's at least one more bug there. And maybe if that is fixed, the condition in this PR doesn't even need `isRef`.

My head is spinning.

Comment (Geod24) [reply src/dmd/expressionsem.d]: ^ Looking at the history, it looks like you figured it out ?

Comment (aG0aep6G) [reply src/dmd/expressionsem.d]: > ^ Looking at the history, it looks like you figured it out ?

No. The tests pass, but I'm still not sure if the condition is sufficient. And I don't think I'm going to figure it out.

**Comment by 12345swordy** [comment]
So what is stopping this from being Pulled?

**Comment by dkorpel** [comment]
This is at odds with [issue 22522 - Creating interior pointers allowed in @safe](https://issues.dlang.org/show_bug.cgi?id=22522).

**Pull request closed without merging**